### PR TITLE
Force off-main-thread ItemsSource updates to update on main thread

### LINKED ISSUE #9753: [Bug, CollectionView] CollectionView crashes when ItemsSource updated too fast
### Description
When CollectionView is bound to a ObservableCollection and it is updated before display fade in animation has finished it will crash.

### Steps to Reproduce

1. Create CollectionView and bind an ObservableCollection to ItemsSource
2. Quickly add several items to ObservableCollection

### Basic Information

- Version with issue: 4.5.0.356
- Last known good version: --
- IDE: VS for Mac
- Platform Target Frameworks: 
  - Android: 9.1
- Android Support Library Version: 10.0

### Exception Log
```
Exception #0: Inconsistency detected. Invalid item position 48(offset:49).state:49 crc643f46942d9dd1fff9.CollectionViewRenderer{ba0eea9 VFED..... ......ID 0,0-394,652 #1d1}, adapter:crc643f46942d9dd1fff9.GroupableItemsViewAdapter_2@16bfd3a, layout:androidx.recyclerview.widget.LinearLayoutManager@3183dc7, context:android.view.ContextThemeWrapper@42bf230

  --- End of managed Java.Lang.IndexOutOfBoundsException stack trace ---
java.lang.IndexOutOfBoundsException: Inconsistency detected. Invalid item position 48(offset:49).state:49 crc643f46942d9dd1fff9.CollectionViewRenderer{ba0eea9 VFED..... ......ID 0,0-394,652 #1d1}, adapter:crc643f46942d9dd1fff9.GroupableItemsViewAdapter_2@16bfd3a, layout:androidx.recyclerview.widget.LinearLayoutManager@3183dc7, context:android.view.ContextThemeWrapper@42bf230
	at androidx.recyclerview.widget.RecyclerView$Recycler.tryGetViewHolderForPositionByDeadline(RecyclerView.java:6183)
	at androidx.recyclerview.widget.RecyclerView$Recycler.getViewForPosition(RecyclerView.java:6118)
	at androidx.recyclerview.widget.RecyclerView$Recycler.getViewForPosition(RecyclerView.java:6114)
	at androidx.recyclerview.widget.LinearLayoutManager$LayoutState.next(LinearLayoutManager.java:2303)
	at androidx.recyclerview.widget.LinearLayoutManager.layoutChunk(LinearLayoutManager.java:1627)
	at androidx.recyclerview.widget.LinearLayoutManager.fill(LinearLayoutManager.java:1587)
	at androidx.recyclerview.widget.LinearLayoutManager.onLayoutChildren(LinearLayoutManager.java:665)
	at androidx.recyclerview.widget.RecyclerView.dispatchLayoutStep1(RecyclerView.java:4085)
	at androidx.recyclerview.widget.RecyclerView.dispatchLayout(RecyclerView.java:3849)
	at androidx.recyclerview.widget.RecyclerView.consumePendingUpdateOperations(RecyclerView.java:1897)
	at androidx.recyclerview.widget.RecyclerView$ViewFlinger.run(RecyclerView.java:5260)
	at android.view.Choreographer$CallbackRecord.run(Choreographer.java:933)
	at android.view.Choreographer.doCallbacks(Choreographer.java:742)
	at android.view.Choreographer.doFrame(Choreographer.java:671)
	at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:919)
	at android.os.Handler.handleCallback(Handler.java:761)
	at android.os.Handler.dispatchMessage(Handler.java:98)
	at android.os.Looper.loop(Looper.java:156)
	at android.app.ActivityThread.main(ActivityThread.java:6623)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:942)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:832)
```

### Workaround
-> Update all CollectionView changing properties on main thread.
-> Throttle elements added to observable collection interval by 250ms.

### LINKED ISSUE #10735: [Bug] [Fatal] [Android] CollectionView Causes Application Crash When Keyboard Opens
### Description

This bug is fatal and not noticeable easily but it will happen in the app sooner or later.

If The CollectionView has **ItemsUpdatingScrollMode="KeepLastItemInView"** It will cause the application to crash when the keyboard opens (or closes not sure) throwing this exception

**Java.Lang.IndexOutOfBoundsException
  Message=Inconsistency detected. Invalid view holder adapter positionViewHolder{41a1e09 position=3 id=-1, oldPos=2, pLpos:2 scrap [attachedScrap] tmpDetached no parent} crc643f46942d9dd1fff9.CollectionViewRenderer{dc8ea8d VFED..... ......ID 0,0-1080,1590 # 2}, adapter:crc643f46942d9dd1fff9.GroupableItemsViewAdapter_2@a2f228e, layout:android.support.v7.widget.LinearLayoutManager@8fed99a, context:android.view.ContextThemeWrapper@1f09f24**

### Steps to Reproduce
* Run the attached solution, It is automated.

### Expected Behavior
CollectionView to keep adding items 

### Actual Behavior
Application Crashes with above exception

### Basic Information
* Android 9 and 10, Pixel simulator and Samsung device
* XF Version: 4.6.0.726

### Workaround
Set ItemsUpdatingScrollMode to other values and handle scrolling by yourself

[CollectionViewException.zip](https://github.com/xamarin/Xamarin.Forms/files/4640103/CollectionViewException.zip)

### PR DESCRIPTION
**Pull request by hartez**
### Description of Change ###

If a developer 

- uses an ObservableCollection (technically, anything which implements INotifyCollectionChange) as the ItemsSource for a CollectionView
- makes changes to that ItemsSource off of the main thread (e.g., using a threadpool thread)
- does all this on Android

for the most part, these changes will be automatically marshaled by ObservableItemsSource to the main thread, and all will be well.

However, if the changes are timed correctly (or occur quickly enough) and there is any scrolling involved (e.g., using KeepLastItemInView, or opening/closing the keyboard, or an ill-timed user scrolling action), it's possible that the RecyclerView will attempt to access the underlying collection in a way which will make the viewholder collection inconsistent. At this point, the user may see errors like:

`java.lang.IndexOutOfBoundsException: Inconsistency detected. Invalid item position`

or

`java.Lang.IndexOutOfBoundsException: Inconsistency detected. Invalid view holder adapter position`

The issue is that the RecyclerView (via the adapter) is accessing the underlying source between the time the changes have been made (from a non-UI thread) and the time the adapter has been notified of the changes (which must happen on the UI thread). 

To work around this problem, these changes provide a wrapper class which queues up the changes requested on the non-ui-thread and later processes them on the UI thread; in the intervening time, any requests for data are answered with a snapshot of the data before it was modified. The wrapper class is injected by the ItemsSourceFactory on Android.

Additionally, these changes address an issue where the ItemsViewRenderer was using the adapter's item count for scrolling rather than the item count in the layout manager. This would occasionally cause issues with the KeepLastItemInView option.

### Issues Resolved ### 

- fixes #10735
- fixes #9753

### API Changes ###

Added to Core:

public class MarshalingObservableCollection

### Platforms Affected ### 

- Android

### Behavioral/Visual Changes ###

None

### Before/After Screenshots ### 

Not applicable

### Testing Procedure ###

Issue 10735 manual test
UI tests
Automated platform tests

### PR Checklist ###

- [ ] Targets the correct branch
- [ ] Tests are passing (or failures are unrelated)

### CODE REVIEW

File: Xamarin.Forms.Core/Items/MarshalingObservableCollection.cs [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+using System.Collections.Specialized;
+
+namespace Xamarin.Forms
+{
+	// Wraps a List which implements INotifyCollectionChanged (usually an ObservableCollection)
+	// and marshals all of the list modifications to the main thread. Modifications to the underlying 
+	// collection which are made off of the main thread remain invisible to consumers on the main thread
+	// until they have been processed by the main thread.
+
+	public class MarshalingObservableCollection : List<object>, INotifyCollectionChanged
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Core/Items/MarshalingObservableCollection.cs:13]: public ?

Comment (hartez) [reply Xamarin.Forms.Core/Items/MarshalingObservableCollection.cs:13]: Used by Platform.Android, so it needs to be public.

File: Xamarin.Forms.Platform.Android/CollectionView/ItemsSourceFactory.cs [hunk 2]
```diff
@@ -21,8 +23,8 @@ public static IItemsViewSource Create(IEnumerable itemsSource, ICollectionChange
 
 			switch (itemsSource)
 			{
-				case IList _ when itemsSource is INotifyCollectionChanged:
-					return new ObservableItemsSource(itemsSource as IList, notifier);
+				case IList list when itemsSource is INotifyCollectionChanged:
+					return new ObservableItemsSource(new MarshalingObservableCollection(list), notifier);
```
Comment (jsuarezruiz) [inline Xamarin.Forms.Platform.Android/CollectionView/ItemsSourceFactory.cs:27]: What about iOS? Would it be necessary to use `MarshalingObservableCollection` in iOS too?

**Comment by Uzivatel919** [comment]
Still having this issue on Xamarin.Forms 4.8.0.1687. Is the scenario with fast updates. Let image custom collection derived from ObservableCollection holding another collection (groups). When these groups are fast replaced there comes error.

```Java.Lang.IndexOutOfBoundsException: 'Inconsistency detected. Invalid item position 15(offset:15).state:16 crc643f46942d9dd1fff9.CollectionViewRenderer{a46a0d7 VFED..... ......ID 0,0-1080,1584 #1a}, adapter:crc643f46942d9dd1fff9.GroupableItemsViewAdapter_2@5ea7373, layout:androidx.recyclerview.widget.LinearLayoutManager@353715c, context:android.view.ContextThemeWrapper@d597856'```

```0xFFFFFFFFFFFFFFFF in System.Diagnostics.Debugger.Mono_UnhandledException_internal	C#
0x1 in System.Diagnostics.Debugger.Mono_UnhandledException at /Users/builder/jenkins/workspace/archive-mono/2020-02/android/release/mcs/class/corlib/System.Diagnostics/Debugger.cs:125,4	C#
0x3E in Android.Runtime.DynamicMethodNameCounter.113	C#
0x12 in System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw at /Users/builder/jenkins/workspace/archive-mono/2020-02/android/release/mcs/class/referencesource/mscorlib/system/runtime/exceptionservices/exceptionservicescommon.cs:157,13	C#
0x8E in Java.Interop.JniEnvironment.InstanceMethods.CallNonvirtualVoidMethod	C#
0x5F in Java.Interop.JniPeerMembers.JniInstanceMethods.InvokeVirtualVoidMethod	C#
0x80 in AndroidX.RecyclerView.Widget.RecyclerView.OnLayout at D:\a\1\s\generated\androidx.recyclerview.recyclerview\obj\Release\monoandroid90\generated\src\AndroidX.RecyclerView.Widget.RecyclerView.cs:13220,5	C#
0x8 in Xamarin.Forms.Platform.Android.ItemsViewRenderer<Xamarin.Forms.GroupableItemsView,Xamarin.Forms.Platform.Android.GroupableItemsViewAdapter<Xamarin.Forms.GroupableItemsView,Xamarin.Forms.Platform.Android.IGroupableItemsViewSource>,Xamarin.Forms.Platform.Android.IGroupableItemsViewSource>.OnLayout at D:\a\1\s\Xamarin.Forms.Platform.Android\CollectionView\ItemsViewRenderer.cs:68,4	C#
0x10 in AndroidX.RecyclerView.Widget.RecyclerView.n_OnLayout_ZIIII at D:\a\1\s\generated\androidx.recyclerview.recyclerview\obj\Release\monoandroid90\generated\src\AndroidX.RecyclerView.Widget.RecyclerView.cs:13204,4	C#
0x2F in Android.Runtime.DynamicMethodNameCounter.113	C#
0xFFFFFFFFFFFFFFFF in Java.Interop.NativeMethods.java_interop_jnienv_call_nonvirtual_void_method_a	C#
0x79 in Java.Interop.JniEnvironment.InstanceMethods.CallNonvirtualVoidMethod	C#
0x5F in Java.Interop.JniPeerMembers.JniInstanceMethods.InvokeVirtualVoidMethod	C#
0x80 in AndroidX.SwipeRefreshLayout.Widget.SwipeRefreshLayout.OnLayout at D:\a\1\s\generated\androidx.swiperefreshlayout.swiperefreshlayout\obj\Release\monoandroid90\generated\src\AndroidX.SwipeRefreshLayout.Widget.SwipeRefreshLayout.cs:539,5	C#
0x8 in Xamarin.Forms.Platform.Android.RefreshViewRenderer.OnLayout at D:\a\1\s\Xamarin.Forms.Platform.Android\Renderers\RefreshViewRenderer.cs:74,4	C#
0x10 in AndroidX.SwipeRefreshLayout.Widget.SwipeRefreshLayout.n_OnLayout_ZIIII at D:\a\1\s\generated\androidx.swiperefreshlayout.swiperefreshlayout\obj\Release\monoandroid90\generated\src\AndroidX.SwipeRefreshLayout.Widget.SwipeRefreshLayout.cs:523,4	C#
0x2F in Android.Runtime.DynamicMethodNameCounter.111	C#
0xFFFFFFFFFFFFFFFF in Java.Interop.NativeMethods.java_interop_jnienv_call_nonvirtual_void_method_a	C#
0x79 in Java.Interop.JniEnvironment.InstanceMethods.CallNonvirtualVoidMethod	C#
0x21 in Java.Interop.JniPeerMembers.JniInstanceMethods.InvokeNonvirtualVoidMethod	C#
0x69 in Android.Views.ViewGroup.Layout	C#
0x118 in Xamarin.Forms.Platform.Android.VisualElementTracker.UpdateLayout at D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementTracker.cs:105,5	C#
0xB in Xamarin.Forms.Platform.Android.RefreshViewRenderer.UpdateLayout at D:\a\1\s\Xamarin.Forms.Platform.Android\Renderers\RefreshViewRenderer.cs:233,33	C#
0x3A in Xamarin.Forms.Platform.Android.VisualElementRenderer<Xamarin.Forms.View>.UpdateLayout at D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementRenderer.cs:422,5	C#
0x1E in Xamarin.Forms.Platform.Android.VisualElementRenderer<Xamarin.Forms.View>.OnLayout at D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementRenderer.cs:400,4	C#
0x8 in Xamarin.Forms.Platform.Android.Platform.DefaultRenderer.OnLayout at D:\a\1\s\Xamarin.Forms.Platform.Android\Platform.cs:1313,5	C#
0x10 in Xamarin.Forms.Platform.Android.FormsViewGroup.n_OnLayout_ZIIII at D:\a\1\s\Xamarin.Forms.Platform.Android.FormsViewGroup\obj\Release\generated\src\Xamarin.Forms.Platform.Android.FormsViewGroup.cs:198,4	C#
0x2F in Android.Runtime.DynamicMethodNameCounter.33	C#
0xFFFFFFFFFFFFFFFF in Java.Interop.NativeMethods.java_interop_jnienv_call_nonvirtual_void_method_a	C#
0x79 in Java.Interop.JniEnvironment.InstanceMethods.CallNonvirtualVoidMethod	C#
0x5F in Java.Interop.JniPeerMembers.JniInstanceMethods.InvokeVirtualVoidMethod	C#
0x97 in Xamarin.Forms.Platform.Android.FormsViewGroup.MeasureAndLayout at D:\a\1\s\Xamarin.Forms.Platform.Android.FormsViewGroup\obj\Release\generated\src\Xamarin.Forms.Platform.Android.FormsViewGroup.cs:181,5	C#
0x171 in Xamarin.Forms.Platform.Android.VisualElementTracker.UpdateLayout at D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementTracker.cs:111,5	C#
0x1E in Xamarin.Forms.Platform.Android.VisualElementRenderer<Xamarin.Forms.View>.UpdateLayout at D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementRenderer.cs:143,4	C#
0x3A in Xamarin.Forms.Platform.Android.VisualElementRenderer<Xamarin.Forms.Page>.UpdateLayout at D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementRenderer.cs:422,5	C#
0x1E in Xamarin.Forms.Platform.Android.VisualElementRenderer<Xamarin.Forms.Page>.OnLayout at D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementRenderer.cs:400,4	C#
0x8 in Xamarin.Forms.Platform.Android.PageRenderer.OnLayout at D:\a\1\s\Xamarin.Forms.Platform.Android\Renderers\PageRenderer.cs:239,4	C#
0x10 in Xamarin.Forms.Platform.Android.FormsViewGroup.n_OnLayout_ZIIII at D:\a\1\s\Xamarin.Forms.Platform.Android.FormsViewGroup\obj\Release\generated\src\Xamarin.Forms.Platform.Android.FormsViewGroup.cs:198,4	C#
0x2F in Android.Runtime.DynamicMethodNameCounter.33	C#
0xFFFFFFFFFFFFFFFF in Java.Interop.NativeMethods.java_interop_jnienv_call_nonvirtual_void_method_a	C#
0x79 in Java.Interop.JniEnvironment.InstanceMethods.CallNonvirtualVoidMethod	C#
0x5F in Java.Interop.JniPeerMembers.JniInstanceMethods.InvokeVirtualVoidMethod	C#
0x97 in Xamarin.Forms.Platform.Android.FormsViewGroup.MeasureAndLayout at D:\a\1\s\Xamarin.Forms.Platform.Android.FormsViewGroup\obj\Release\generated\src\Xamarin.Forms.Platform.Android.FormsViewGroup.cs:181,5	C#
0x171 in Xamarin.Forms.Platform.Android.VisualElementTracker.UpdateLayout at D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementTracker.cs:111,5	C#
0x1E in Xamarin.Forms.Platform.Android.VisualElementRenderer<Xamarin.Forms.Page>.UpdateLayout at D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementRenderer.cs:143,4	C#
0x6 in Xamarin.Forms.Platform.Android.PageContainer.OnLayout at D:\a\1\s\Xamarin.Forms.Platform.Android\Renderers\PageContainer.cs:30,4	C#
0x53 in Xamarin.Forms.Platform.Android.ShellPageContainer.OnLayout at D:\a\1\s\Xamarin.Forms.Platform.Android\Renderers\ShellPageContainer.cs:16,4	C#
0x10 in Android.Views.ViewGroup.n_OnLayout_ZIIII	C#
0x2F in Android.Runtime.DynamicMethodNameCounter.27	C#
0xFFFFFFFFFFFFFFFF in Java.Interop.NativeMethods.java_interop_jnienv_call_nonvirtual_void_method_a	C#
0x79 in Java.Interop.JniEnvironment.InstanceMethods.CallNonvirtualVoidMethod	C#
0x21 in Java.Interop.JniPeerMembers.JniInstanceMethods.InvokeNonvirtualVoidMethod	C#
0x69 in Android.Views.ViewGroup.Layout	C#
0x3F in Xamarin.Forms.Platform.Android.ShellRenderer.UpdateLayout at D:\a\1\s\Xamarin.Forms.Platform.Android\Renderers\ShellRenderer.cs:83,4	C#
0x2B in Xamarin.Forms.Platform.Android.AppCompat.Platform.Xamarin.Forms.Platform.Android.IPlatformLayout.OnLayout at D:\a\1\s\Xamarin.Forms.Platform.Android\AppCompat\Platform.cs:235,4	C#
0x2D in Xamarin.Forms.Platform.Android.PlatformRenderer.OnLayout at D:\a\1\s\Xamarin.Forms.Platform.Android\PlatformRenderer.cs:75,4	C#
0x10 in Android.Views.ViewGroup.n_OnLayout_ZIIII	C#
0x2F in Android.Runtime.DynamicMethodNameCounter.27	C#```

**Comment by Uzivatel919** [comment]
Still having this issue. `CollectionView` has `ItemsUpdatingScrollMode="KeepItemsInView"`. Using the `ObservableCollection` for `ItemsSource`. Xamarin.Forms 4.8.0.1687.

```Java.Lang.IndexOutOfBoundsException: 'Inconsistency detected. Invalid item position 17(offset:17).state:18 crc643f46942d9dd1fff9.CollectionViewRenderer{c1f8419 VFED..... ......ID 0,0-1080,1584 #1a}, adapter:crc643f46942d9dd1fff9.GroupableItemsViewAdapter_2@d9b48d5, layout:androidx.recyclerview.widget.LinearLayoutManager@d1bd9b6, context:android.view.ContextThemeWrapper@73ab760'```

```0xFFFFFFFFFFFFFFFF in System.Diagnostics.Debugger.Mono_UnhandledException_internal	C#
0x1 in System.Diagnostics.Debugger.Mono_UnhandledException at /Users/builder/jenkins/workspace/archive-mono/2020-02/android/release/mcs/class/corlib/System.Diagnostics/Debugger.cs:125,4	C#
0x3E in Android.Runtime.DynamicMethodNameCounter.113	C#
0x12 in System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw at /Users/builder/jenkins/workspace/archive-mono/2020-02/android/release/mcs/class/referencesource/mscorlib/system/runtime/exceptionservices/exceptionservicescommon.cs:157,13	C#
0x8E in Java.Interop.JniEnvironment.InstanceMethods.CallNonvirtualVoidMethod	C#
0x5F in Java.Interop.JniPeerMembers.JniInstanceMethods.InvokeVirtualVoidMethod	C#
0x80 in AndroidX.RecyclerView.Widget.RecyclerView.OnLayout at D:\a\1\s\generated\androidx.recyclerview.recyclerview\obj\Release\monoandroid90\generated\src\AndroidX.RecyclerView.Widget.RecyclerView.cs:13220,5	C#
0x8 in Xamarin.Forms.Platform.Android.ItemsViewRenderer<Xamarin.Forms.GroupableItemsView,Xamarin.Forms.Platform.Android.GroupableItemsViewAdapter<Xamarin.Forms.GroupableItemsView,Xamarin.Forms.Platform.Android.IGroupableItemsViewSource>,Xamarin.Forms.Platform.Android.IGroupableItemsViewSource>.OnLayout at D:\a\1\s\Xamarin.Forms.Platform.Android\CollectionView\ItemsViewRenderer.cs:68,4	C#
0x10 in AndroidX.RecyclerView.Widget.RecyclerView.n_OnLayout_ZIIII at D:\a\1\s\generated\androidx.recyclerview.recyclerview\obj\Release\monoandroid90\generated\src\AndroidX.RecyclerView.Widget.RecyclerView.cs:13204,4	C#
0x2F in Android.Runtime.DynamicMethodNameCounter.113	C#
0xFFFFFFFFFFFFFFFF in Java.Interop.NativeMethods.java_interop_jnienv_call_nonvirtual_void_method_a	C#
0x79 in Java.Interop.JniEnvironment.InstanceMethods.CallNonvirtualVoidMethod	C#
0x5F in Java.Interop.JniPeerMembers.JniInstanceMethods.InvokeVirtualVoidMethod	C#
0x80 in AndroidX.SwipeRefreshLayout.Widget.SwipeRefreshLayout.OnLayout at D:\a\1\s\generated\androidx.swiperefreshlayout.swiperefreshlayout\obj\Release\monoandroid90\generated\src\AndroidX.SwipeRefreshLayout.Widget.SwipeRefreshLayout.cs:539,5	C#
0x8 in Xamarin.Forms.Platform.Android.RefreshViewRenderer.OnLayout at D:\a\1\s\Xamarin.Forms.Platform.Android\Renderers\RefreshViewRenderer.cs:74,4	C#
0x10 in AndroidX.SwipeRefreshLayout.Widget.SwipeRefreshLayout.n_OnLayout_ZIIII at D:\a\1\s\generated\androidx.swiperefreshlayout.swiperefreshlayout\obj\Release\monoandroid90\generated\src\AndroidX.SwipeRefreshLayout.Widget.SwipeRefreshLayout.cs:523,4	C#
0x2F in Android.Runtime.DynamicMethodNameCounter.111	C#
0xFFFFFFFFFFFFFFFF in Java.Interop.NativeMethods.java_interop_jnienv_call_nonvirtual_void_method_a	C#
0x79 in Java.Interop.JniEnvironment.InstanceMethods.CallNonvirtualVoidMethod	C#
0x21 in Java.Interop.JniPeerMembers.JniInstanceMethods.InvokeNonvirtualVoidMethod	C#
0x69 in Android.Views.ViewGroup.Layout	C#
0x118 in Xamarin.Forms.Platform.Android.VisualElementTracker.UpdateLayout at D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementTracker.cs:105,5	C#
0xB in Xamarin.Forms.Platform.Android.RefreshViewRenderer.UpdateLayout at D:\a\1\s\Xamarin.Forms.Platform.Android\Renderers\RefreshViewRenderer.cs:233,33	C#
0x3A in Xamarin.Forms.Platform.Android.VisualElementRenderer<Xamarin.Forms.View>.UpdateLayout at D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementRenderer.cs:422,5	C#
0x1E in Xamarin.Forms.Platform.Android.VisualElementRenderer<Xamarin.Forms.View>.OnLayout at D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementRenderer.cs:400,4	C#
0x8 in Xamarin.Forms.Platform.Android.Platform.DefaultRenderer.OnLayout at D:\a\1\s\Xamarin.Forms.Platform.Android\Platform.cs:1313,5	C#
0x10 in Xamarin.Forms.Platform.Android.FormsViewGroup.n_OnLayout_ZIIII at D:\a\1\s\Xamarin.Forms.Platform.Android.FormsViewGroup\obj\Release\generated\src\Xamarin.Forms.Platform.Android.FormsViewGroup.cs:198,4	C#
0x2F in Android.Runtime.DynamicMethodNameCounter.33	C#
0xFFFFFFFFFFFFFFFF in Java.Interop.NativeMethods.java_interop_jnienv_call_nonvirtual_void_method_a	C#
0x79 in Java.Interop.JniEnvironment.InstanceMethods.CallNonvirtualVoidMethod	C#
0x5F in Java.Interop.JniPeerMembers.JniInstanceMethods.InvokeVirtualVoidMethod	C#
0x97 in Xamarin.Forms.Platform.Android.FormsViewGroup.MeasureAndLayout at D:\a\1\s\Xamarin.Forms.Platform.Android.FormsViewGroup\obj\Release\generated\src\Xamarin.Forms.Platform.Android.FormsViewGroup.cs:181,5	C#
0x171 in Xamarin.Forms.Platform.Android.VisualElementTracker.UpdateLayout at D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementTracker.cs:111,5	C#
0x1E in Xamarin.Forms.Platform.Android.VisualElementRenderer<Xamarin.Forms.View>.UpdateLayout at D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementRenderer.cs:143,4	C#
0x3A in Xamarin.Forms.Platform.Android.VisualElementRenderer<Xamarin.Forms.Page>.UpdateLayout at D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementRenderer.cs:422,5	C#
0x1E in Xamarin.Forms.Platform.Android.VisualElementRenderer<Xamarin.Forms.Page>.OnLayout at D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementRenderer.cs:400,4	C#
0x8 in Xamarin.Forms.Platform.Android.PageRenderer.OnLayout at D:\a\1\s\Xamarin.Forms.Platform.Android\Renderers\PageRenderer.cs:239,4	C#
0x10 in Xamarin.Forms.Platform.Android.FormsViewGroup.n_OnLayout_ZIIII at D:\a\1\s\Xamarin.Forms.Platform.Android.FormsViewGroup\obj\Release\generated\src\Xamarin.Forms.Platform.Android.FormsViewGroup.cs:198,4	C#
0x2F in Android.Runtime.DynamicMethodNameCounter.33	C#
0xFFFFFFFFFFFFFFFF in Java.Interop.NativeMethods.java_interop_jnienv_call_nonvirtual_void_method_a	C#
0x79 in Java.Interop.JniEnvironment.InstanceMethods.CallNonvirtualVoidMethod	C#
0x5F in Java.Interop.JniPeerMembers.JniInstanceMethods.InvokeVirtualVoidMethod	C#
0x97 in Xamarin.Forms.Platform.Android.FormsViewGroup.MeasureAndLayout at D:\a\1\s\Xamarin.Forms.Platform.Android.FormsViewGroup\obj\Release\generated\src\Xamarin.Forms.Platform.Android.FormsViewGroup.cs:181,5	C#
0x171 in Xamarin.Forms.Platform.Android.VisualElementTracker.UpdateLayout at D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementTracker.cs:111,5	C#
0x1E in Xamarin.Forms.Platform.Android.VisualElementRenderer<Xamarin.Forms.Page>.UpdateLayout at D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementRenderer.cs:143,4	C#
0x6 in Xamarin.Forms.Platform.Android.PageContainer.OnLayout at D:\a\1\s\Xamarin.Forms.Platform.Android\Renderers\PageContainer.cs:30,4	C#
0x53 in Xamarin.Forms.Platform.Android.ShellPageContainer.OnLayout at D:\a\1\s\Xamarin.Forms.Platform.Android\Renderers\ShellPageContainer.cs:16,4	C#
0x10 in Android.Views.ViewGroup.n_OnLayout_ZIIII	C#
0x2F in Android.Runtime.DynamicMethodNameCounter.27	C#
0xFFFFFFFFFFFFFFFF in Java.Interop.NativeMethods.java_interop_jnienv_call_nonvirtual_void_method_a	C#
0x79 in Java.Interop.JniEnvironment.InstanceMethods.CallNonvirtualVoidMethod	C#
0x21 in Java.Interop.JniPeerMembers.JniInstanceMethods.InvokeNonvirtualVoidMethod	C#
0x69 in Android.Views.ViewGroup.Layout	C#
0x3F in Xamarin.Forms.Platform.Android.ShellRenderer.UpdateLayout at D:\a\1\s\Xamarin.Forms.Platform.Android\Renderers\ShellRenderer.cs:83,4	C#
0x2B in Xamarin.Forms.Platform.Android.AppCompat.Platform.Xamarin.Forms.Platform.Android.IPlatformLayout.OnLayout at D:\a\1\s\Xamarin.Forms.Platform.Android\AppCompat\Platform.cs:235,4	C#
0x2D in Xamarin.Forms.Platform.Android.PlatformRenderer.OnLayout at D:\a\1\s\Xamarin.Forms.Platform.Android\PlatformRenderer.cs:75,4	C#
0x10 in Android.Views.ViewGroup.n_OnLayout_ZIIII	C#
0x2F in Android.Runtime.DynamicMethodNameCounter.27	C#```

**Comment by hartez** [comment]
> My case. ObservableCollection holds collections. These ones are replaced – one by one.

Are you saying you have something like `ObservableCollection<ObservableCollection<object>>`?

**Comment by Uzivatel919** [comment]
> Are you saying you have something like `ObservableCollection<ObservableCollection<object>>`?

No, no. Let review the design. My collegue told me that test regard this issue covers just `Add` method. Maybe there is another issue when replacing. 

```internal class CustomObservableCollection : ObservableCollection<CustomCollection>
{
	public void Replace(CustomCollection col, int index) => SetItem(index, col);
	
	protected override void OnCollectionChanged(NotifyCollectionChangedEventArgs e) 
	{ 
	    BeginInvokeOnMainThread(() => base.OnCollectionChanged(e)); 
	} 
}

internal class CustomCollection: IReadOnlyCollection<Dto> { … }
```

Issue also happens when code is refactored to

```
…
 BeginInvokeOnMainThread(() => _customObservableCollection,Replace(col, index));
…

internal class CustomObservableCollection : ObservableCollection<CustomCollection>
{
	public void Replace(CustomCollection col, int index) => SetItem(index, col);
	
	protected override void OnCollectionChanged(NotifyCollectionChangedEventArgs e) 
	{ 
	    base.OnCollectionChanged(e); 
	} 
}

**Pull request merged**